File: tests/mobility_test.cpp
/**
 * @file mobility_test.cpp
 * @brief ENIGMA Mobility module demonstration.
 *
 * Shows how to:
 *  - Declare the coords directory as a zone property in the XML
 *  - Load all traces via MobilityManager
 *  - Query device positions inside SimGrid actors
 *  - Dump snapshots to JSON + CSV for post-simulation visualisation
 *
 * Usage
 * -----
 *   ./mobility_test_app <platform_file.xml> [output_prefix]
 *
 * The platform XML must contain either a zone- or host-level property:
 *   <prop id="mobility_dir" value="platforms/coords/"/>
 *
 * Alternatively pass --mobility-dir <dir> as argument (see below).
 */

#include <simgrid/s4u.hpp>
#include <iostream>

#include "mobility/MobilityManager.hpp"
#include "mobility/MobilityPosition.hpp"

XBT_LOG_NEW_DEFAULT_CATEGORY(mobility_test, "ENIGMA Mobility Test");

namespace sg4 = simgrid::s4u;
using namespace enigma::mobility;

// -------------------------------------------------------------------------- //
// Actor: periodically logs its own position to the SimGrid log              //
// -------------------------------------------------------------------------- //

class MobileActor {
    std::string    host_name_;
    MobilityManager* mob_;
    double         interval_;
    int            iterations_;

public:
    MobileActor(std::string name, MobilityManager* mob,
                double interval = 2.0, int iterations = 10)
        : host_name_(std::move(name))
        , mob_(mob)
        , interval_(interval)
        , iterations_(iterations) {}

    void operator()() const {
        for (int i = 0; i < iterations_; ++i) {
            double t = sg4::Engine::get_clock();
            auto pos = mob_->position_at(host_name_, t);
            if (pos) {
                double spd = pos->extra.count("speed")   ? pos->extra.at("speed")   : 0.0;
                double hdg = pos->extra.count("heading") ? pos->extra.at("heading") : 0.0;
                XBT_INFO("[%.2f s] %s → lat=%.6f lon=%.6f  spd=%.1f m/s  hdg=%.1f°",
                         t, host_name_.c_str(),
                         pos->latitude, pos->longitude, spd, hdg);
            } else {
                XBT_INFO("[%.2f s] %s → no mobility trace available", t, host_name_.c_str());
            }
            // Simulate some local work
            sg4::this_actor::execute(5e8);
            sg4::this_actor::sleep_for(interval_);
        }
        XBT_INFO("%s: actor finished", host_name_.c_str());
    }
};

// -------------------------------------------------------------------------- //
// main                                                                       //
// -------------------------------------------------------------------------- //

int main(int argc, char** argv) {
    sg4::Engine e(&argc, argv);

    if (argc < 2) {
        std::cerr << "Usage: " << argv[0]
                  << " <platform.xml> [output_prefix] [--mobility-dir <dir>]\n";
        return 1;
    }

    const std::string platform_file = argv[1];
    std::string output_prefix = "mobility_output";
    std::string mobility_dir;

    for (int i = 2; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--mobility-dir" && i + 1 < argc) {
            mobility_dir = argv[++i];
        } else if (arg[0] != '-') {
            output_prefix = arg;
        }
    }

    e.load_platform(platform_file);

    // ------------------------------------------------------------------ //
    // Create MobilityManager (reads mobility_dir from XML prop or arg)    //
    // ------------------------------------------------------------------ //
    std::unique_ptr<MobilityManager> mob;
    if (!mobility_dir.empty()) {
        mob = std::make_unique<MobilityManager>(e, mobility_dir);
    } else {
        mob = std::make_unique<MobilityManager>(e);
    }

    XBT_INFO("=== Mobility Test Application ===");
    XBT_INFO("Platform file : %s", platform_file.c_str());
    XBT_INFO("Traces loaded : %zu", mob->trace_count());
    XBT_INFO("Output prefix : %s", output_prefix.c_str());

    // ------------------------------------------------------------------ //
    // Start periodic snapshot actor (records positions every 0.5 sim s)   //
    // ------------------------------------------------------------------ //
    mob->start_periodic_actor(e, 0.5);

    // ------------------------------------------------------------------ //
    // Deploy a MobileActor on every host that has a trace                 //
    // ------------------------------------------------------------------ //
    auto hosts = e.get_all_hosts();
    int deployed = 0;
    for (auto* host : hosts) {
        if (mob->has_trace(host->get_cname())) {
            host->add_actor("mobile_actor",
                            MobileActor(host->get_cname(), mob.get(), 1.0, 8));
            XBT_INFO("Deployed mobile actor on '%s'", host->get_cname());
            ++deployed;
        }
    }

    if (deployed == 0) {
        XBT_WARN("No matching traces found for any host. "
                 "Check that CSV filenames match host names.");
    }

    // ------------------------------------------------------------------ //
    // Run simulation                                                       //
    // ------------------------------------------------------------------ //
    e.run();

    XBT_INFO("=== Simulation completed – t=%.3f s ===", sg4::Engine::get_clock());

    // ------------------------------------------------------------------ //
    // Export results                                                       //
    // ------------------------------------------------------------------ //
    const std::string json_out = output_prefix + "_snapshots.json";
    const std::string csv_out  = output_prefix + "_snapshots.csv";
    const std::string raw_json = output_prefix + "_raw_traces.json";

    mob->export_json(json_out);
    mob->export_csv(csv_out);
    mob->export_traces_json(raw_json);

    XBT_INFO("Outputs written:");
    XBT_INFO("  Snapshots JSON : %s  (%zu records)", json_out.c_str(), mob->snapshots().size());
    XBT_INFO("  Snapshots CSV  : %s", csv_out.c_str());
    XBT_INFO("  Raw traces JSON: %s", raw_json.c_str());
    XBT_INFO("Visualise with:");
    XBT_INFO("  python3 src/python/tools/mobility_viewer.py %s", json_out.c_str());

    return 0;
}

File: src/python/tests/mobility_test.py
"""
mobility_test.py
----------------
ENIGMA mobility module demonstration using SimGrid Python bindings.

Shows:
  - Loading device traces from a coords directory (specified via XML property
    OR -coords-dir argument)
  - Querying interpolated positions inside SimGrid actors
  - Live window opened during simulation
  - Recording all snapshots to CSV / JSON / GeoJSON after simulation
  - Animated replay and static map generation

Usage
-----
    python3 tests/mobility_test.py <platform.xml> [options]

    Options:
      --coords-dir <dir>   Override the mobility_dir XML property
      --output <prefix>    Output file prefix (default: mobility_output/)
      --interval <secs>    Snapshot interval in sim-seconds (default: 0.5)
      --no-live            Disable live window (record + replay only)
      --fps <n>            Replay animation fps (default: 15)
      --save-replay        Save replay to mobility_output/replay.gif
"""
from __future__ import annotations

import os
import sys

_HERE = os.path.dirname(os.path.abspath(__file__))
_ROOT = os.path.join(_HERE, "..")
if _ROOT not in sys.path:
    sys.path.insert(0, _ROOT)

import simgrid

from enigma.mobility import (
    MobilityManager,
    MobilityVisualizer,
)


# ─────────────────────────────────────────────────────────────────────────── #
# Actors                                                                       #
# ─────────────────────────────────────────────────────────────────────────── #

class MobileActor:
    """
    A demo SimGrid actor that:
      - Periodically queries its own GPS position from MobilityManager
      - Does a small computation to advance simulated time
      - Logs its coordinates
    """

    def __init__(
        self,
        host_name: str,
        mob: MobilityManager,
        interval_s: float = 1.0,
        iterations: int   = 10,
    ) -> None:
        self._host_name  = host_name
        self._mob        = mob
        self._interval   = interval_s
        self._iterations = iterations

    def __call__(self) -> None:
        simgrid.this_actor.info(
            f"[{self._host_name}] Mobile actor started"
        )
        for i in range(self._iterations):
            t   = simgrid.Engine.clock
            pos = self._mob.record(self._host_name, t)

            if pos:
                spd = pos.extra.get("speed",   0.0)
                hdg = pos.extra.get("heading", 0.0)
                simgrid.this_actor.info(
                    f"[{self._host_name}] t={t:.2f}s  "
                    f"lat={pos.latitude:.6f}  lon={pos.longitude:.6f}  "
                    f"spd={spd:.1f} m/s  hdg={hdg:.0f}°"
                )
            else:
                simgrid.this_actor.info(
                    f"[{self._host_name}] t={t:.2f}s  (no mobility trace)"
                )

            # Light local computation to advance simulated time
            simgrid.this_actor.execute(2e8)
            simgrid.this_actor.sleep_for(self._interval)

        simgrid.this_actor.info(f"[{self._host_name}] Actor finished")


# ─────────────────────────────────────────────────────────────────────────── #
# main                                                                         #
# ─────────────────────────────────────────────────────────────────────────── #

def main() -> None:
    # ---- argument parsing ------------------------------------------------- #
    e = simgrid.Engine(sys.argv)

    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <platform.xml> [--coords-dir <dir>] "
              "[--output <prefix>] [--interval <s>] [--no-live] "
              "[--no-replay] [--fps <n>] [--save-replay]", file=sys.stderr)
        sys.exit(1)

    platform_xml = sys.argv[1]
    coords_dir   = None
    # Default output next to this script, not the working directory
    output_dir   = os.path.join(os.path.dirname(os.path.abspath(__file__)), "mobility_output")
    interval_s   = 0.5
    live_viz     = True
    fps          = 15.0
    save_replay  = False
    show_replay  = True
    offline_mode = False

    i = 2
    while i < len(sys.argv):
        arg = sys.argv[i]
        if arg == "--coords-dir" and i + 1 < len(sys.argv):
            coords_dir = sys.argv[i + 1]; i += 2
        elif arg == "--output" and i + 1 < len(sys.argv):
            output_dir = sys.argv[i + 1]; i += 2
        elif arg == "--interval" and i + 1 < len(sys.argv):
            interval_s = float(sys.argv[i + 1]); i += 2
        elif arg == "--fps" and i + 1 < len(sys.argv):
            fps = float(sys.argv[i + 1]); i += 2
        elif arg == "--no-live":
            live_viz = False; i += 1
        elif arg == "--save-replay":
            save_replay = True; i += 1
        elif arg == "--no-replay":
            show_replay = False; i += 1
        elif arg == "--offline":
            offline_mode = True; i += 1
        elif arg == "--online":
            offline_mode = False; i += 1
        else:
            i += 1

    os.makedirs(output_dir, exist_ok=True)

    # ---- load platform ---------------------------------------------------- #
    e.load_platform(platform_xml)
    hosts = e.all_hosts

    simgrid.this_actor.info("=== ENIGMA Mobility Test Application ===")
    simgrid.this_actor.info(f"Platform  : {platform_xml}")
    simgrid.this_actor.info(f"Hosts     : {len(hosts)}")
    simgrid.this_actor.info(f"Output    : {output_dir}/")

    # ---- visualizer ------------------------------------------------------- #
    viz = None
    if live_viz:
        viz = MobilityVisualizer(
            title="ENIGMA – Device Mobility (live)",
            live_path=os.path.join(output_dir, "mobility_live.html"),
        )
        viz.start()
        simgrid.this_actor.info("Live visualization window opened")

    # ---- mobility manager ------------------------------------------------- #
    mob = MobilityManager(e, coords_dir=coords_dir, visualizer=viz,
                          platform_xml=platform_xml)
    simgrid.this_actor.info(f"Traces loaded : {mob.trace_count}")

    # ---- periodic snapshot actor ------------------------------------------ #
    mob.start_periodic_actor(e, interval_s=interval_s)

    # ---- deploy mobile actors --------------------------------------------- #
    deployed = 0
    for host in hosts:
        if mob.has_trace(host.name):
            simgrid.Actor.create(
                "mobile_actor", host,
                MobileActor(host.name, mob, interval_s=1.5, iterations=8),
            )
            deployed += 1
            simgrid.this_actor.info(f"Deployed mobile actor on '{host.name}'")

    if deployed == 0:
        simgrid.this_actor.info(
            "No matching traces found for any host – "
            "check that CSV filenames match host names."
        )

    # ---- run simulation --------------------------------------------------- #
    e.run()

    simgrid.this_actor.info(
        f"=== Simulation completed – t={simgrid.Engine.clock:.3f} s ==="
    )
    simgrid.this_actor.info(
        f"Recorded {mob.recorder.count} snapshots  "
        f"for {len(mob.recorder.devices())} devices"
    )

    mob.recorder.print_stats()

    # ---- stop live window ------------------------------------------------- #
    if viz:
        viz.stop()

    # ---- export recordings ------------------------------------------------ #
    csv_path     = os.path.join(output_dir, "snapshots.csv")
    json_path    = os.path.join(output_dir, "snapshots.json")
    geojson_path = os.path.join(output_dir, "trajectories.geojson")
    map_path     = os.path.join(output_dir, "trajectory_map.html")

    mob.recorder.export_csv(csv_path)
    mob.recorder.export_json(json_path)
    mob.recorder.export_geojson(geojson_path)

    # ---- interactive map -------------------------------------------------- #
    MobilityVisualizer.save_interactive_map(
        mob.recorder, map_path,
        title="ENIGMA – Device Mobility",
        interval_s=interval_s,
        offline=offline_mode,
    )

    simgrid.this_actor.info("Outputs:")
    simgrid.this_actor.info(f"  {csv_path}")
    simgrid.this_actor.info(f"  {json_path}")
    simgrid.this_actor.info(f"  {geojson_path}")
    simgrid.this_actor.info(f"  {map_path}")

    # ---- interactive replay ----------------------------------------------- #
    if show_replay:
        if viz is None:
            viz = MobilityVisualizer(title="ENIGMA – Device Mobility",
                                     live_path=os.path.join(output_dir, "mobility_live.html"),
                                     offline=offline_mode)
        # Always save replay into output_dir (--save-replay controls whether the
        # file is kept; here we always write it so the path stays out of /tmp)
        replay_save = os.path.join(output_dir, "replay.html")
        simgrid.this_actor.info("Opening interactive replay in browser…")
        viz.replay_from_recorder(
            mob.recorder, save_path=replay_save, open_browser=True, offline=offline_mode
        )
    else:
        simgrid.this_actor.info("Replay skipped (--no-replay)")


if __name__ == "__main__":
    main()
